### PR TITLE
feat: template type and configuration fields for template default params and schema

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
**Issue #, if available:**
[P49759307] Add needed structures to EvergreenComponentsCommon repos
**Description of changes:**
Add `aws.greengrass.template` type. Add `TemplateParameter` structure and add `defaultParameters`, `parameterSchema` field to `ComponentConfiguration`.

**Why is this change necessary:**
Necessary for cleaner logic for recipe templating

**How was this change tested:**
Only deserialization tested, since the new features should never be written to a file, only consumed by a deployment.

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java [hunk 1]
```diff
@@ -22,6 +24,14 @@
 
     JsonNode defaultConfiguration;
 
+    JsonNode defaultParameters;
+
+    Map<String, TemplateParameter> parameterSchema;
+
+    public ComponentConfiguration(JsonNode defaultConfiguration) {
+        this.defaultConfiguration = defaultConfiguration;
+    }
```
Comment (MikeDombo) [inline src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: why is this added?

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonDeserialize(builder = TemplateParameter.TemplateParameterBuilder.class)
+public class TemplateParameter {
```
Comment (MikeDombo) [inline src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java:21]: A parameter needs a default too, right? Seems to make more sense to me to put the default in this section instead of a separate section.

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java (hunk 1 above)
Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: the thought is to not break any existing applications that use the `AllArgsConstructor` to create an object

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java (hunk 2 above)
Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java:21]: this is a point I'd like to discuss. I don't think there's a case where a template has configuration, since everything is being handled by the transformer jar. But for readability from the end-user's perspective I think it's useful to separate very clearly the concepts of default config for components and default config for template parameters. One is for the component itself and one is for components other than itself.

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java (hunk 1 above)
Comment (MikeDombo) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: Does anything use allargs?

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java (hunk 2 above)
Comment (MikeDombo) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java:21]: How are they not separate? Putting the defaults in with the schema is much more clear IMO than a separate section.

Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/TemplateParameter.java:21]: Oh I see what you're saying. Sure, that's reasonable. I'll do that.

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java (hunk 1 above)
Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: I have no idea. But at the same time, why would we have `NoArgs` and `AllArgs` if we never expect anyone to use them?

Comment (MikeDombo) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: They exist for Jackson. Jackson needs them both.

Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: I see. But this doesn't address the question of constructor usage. You probably know better than me: is it 100% guaranteed that between the time this PR is merged into my fork and the time it makes it into the main repo, nobody will try to use construct a ComponentConfiguration without using the builder?

Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: I suspect the answer is no, which is why I'm hesitant

Comment (MikeDombo) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: No, but we can fix it as needed. We always use builders, so it is highly highly unlikely that this will be an issue. No one constructs these objects manually anyway, it is always Jackson.

Comment (Victor-C-Zhang) [reply src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentConfiguration.java]: Alright. Good to know

File: src/test/resources/recipes/sample-template-with-all-fields.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
+{
+  "RecipeFormatVersion": "2020-01-25",
+  "ComponentName": "FooTemplate",
+  "ComponentDescription": "Test template for Greengrass packages",
+  "ComponentPublisher": "Me",
+  "ComponentVersion": "1.0.0",
+  "ComponentType": "aws.greengrass.template",
+  "ComponentConfiguration": {
+    "ParameterSchema": {
```
Comment (MikeDombo) [inline src/test/resources/recipes/sample-template-with-all-fields.json]: why is this under component configuration anyhow? Schema isn't config really.... I think this should maybe be a different top-level key.

Comment (Victor-C-Zhang) [reply src/test/resources/recipes/sample-template-with-all-fields.json]: Semantically, we need a `Parameters` field for `ComponentConfiguration` for parameter files, since they can't be modified like normal config K/Vs but also are config values. So either we put both `Parameters` and `ParameterSchema` as top-level keys or we stick both in `ComponentConfiguration`

Comment (MikeDombo) [reply src/test/resources/recipes/sample-template-with-all-fields.json]: Well I'd want the schema taken out from the user input completely

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentRecipe.java [hunk 4]
```diff
@@ -50,6 +50,10 @@
 
     ComponentConfiguration componentConfiguration;
 
+    Map<String, TemplateParameter> parameterSchema;
+
+    Map<String, Object> parameters;
```
Comment (MikeDombo) [inline src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentRecipe.java]: rename `templateParameters`

File: src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentRecipe.java [hunk 5]
```diff
@@ -50,6 +50,10 @@
 
     ComponentConfiguration componentConfiguration;
 
+    Map<String, TemplateParameter> parameterSchema;
```
Comment (MikeDombo) [inline src/main/java/com/amazon/aws/iot/greengrass/component/common/ComponentRecipe.java]: Rename `templateParameterSchema`

**Review by MikeDombo — APPROVED** [review]

**Review by fahadmohammed01 — APPROVED** [review]
